### PR TITLE
[gatsby-source-filesystem] Add AllFiles to directory queries #3727

### PR DESCRIPTION
**Pull request by ptb**
- Added AllFiles query to AllDirectory and directory GraphQL queries.
- Re-organized and simplified this plugin, removing duplicate reassignments and several temporary variables.
- Alphabetized object keys and required packages where possible.
- Updated dependencies.
- Simplified and streamlined file/directory changes queue.
- Simplified wording of reporter messages and comments (using Apple's style guide).

I hope this type of massive re-writing is welcome. This is my first foray into a source plugin and basically had to deconstruct the original to understand how it works. Also, I've never used async/await before, but I think I understand them now. Along the way, I noticed some ways of simplifying assignments as well as adding the functionality requested in issue #3727.

### CODE REVIEW

File: packages/gatsby-source-filesystem/package.json [hunk 1]
```diff
@@ -1,37 +1,41 @@
 {
-  "name": "gatsby-source-filesystem",
-  "description": "Gatsby plugin which parses files within a directory for further parsing by other plugins",
-  "version": "1.5.29",
   "author": "Kyle Mathews <mathews.kyle@gmail.com>",
```
Comment (KyleAMathews) [inline packages/gatsby-source-filesystem/package.json]: We use https://www.npmjs.com/package/fixpack for ordering our package.json files.

File: packages/gatsby-source-filesystem/src/create-file-node.js [hunk 2]
```diff
@@ -1,82 +1,75 @@
...
-      type: `File`,
-    }
+  node = {
+    ...node,
+    allFiles: stats.isDirectory()
+      ? await fs
+          .readdir(node.absolutePath)
+          .map(file => path.join(node.absolutePath, file))
+          .filter(file => fs.statSync(file).isFile())
+          .map(file => createFileNode(file, opts))
```
Comment (KyleAMathews) [inline packages/gatsby-source-filesystem/src/create-file-node.js]: Instead of creating new file duplicate file nodes — you should instead link to them using the `___node` syntax.

Comment (ptb) [reply packages/gatsby-source-filesystem/src/create-file-node.js]: Is there documentation for the ___node syntax? I haven’t come across any mention of that before now.

**Comment by ptb** [comment]
I’ll definitely rework the PR off of the v2 branch. Just glancing at it, the changes there look awesome!

**Comment by ptb** [comment]
I think this code is ready for a second set of eyes. I tried to include the intention of the original code while simplifying the flow and conditionals. There is only one queue now and the queue is flushed as often as circumstances permit.

File: packages/gatsby-source-filesystem/src/create-file-node.js [hunk 3]
```diff
@@ -1,79 +1,74 @@
...
-      accessTime: stats.atime,
-      changeTime: stats.ctime,
-      birthTime: stats.birthtime,
-      ...slashedFile,
-      ...stats,
-    })
-  )
+  node = {
+    ...node,
+    allFile: {
```
Comment (KyleAMathews) [inline packages/gatsby-source-filesystem/src/create-file-node.js]: let's call this `files` as `allFile` suggests it'll behave the same as the top-level field `allFile` — which it won't as this is a simple array without filtering/sorting/etc. options.

File: packages/gatsby-source-filesystem/src/create-file-node.js [hunk 4]
```diff
@@ -1,79 +1,74 @@
...
+      edges: stats.isDirectory()
+        ? await fs
+            .readdir(node.absolutePath)
```
Comment (KyleAMathews) [inline packages/gatsby-source-filesystem/src/create-file-node.js]: I wonder how expensive this is? If it's really cheap, it's fine but we already have read all this data so it could get expensive to double-read information all the way through a deep file tree.

File: packages/gatsby-source-filesystem/src/gatsby-node.js [hunk 5]
```diff
@@ -11,218 +11,159 @@ const { createId, createFileNode } = require(`./create-file-node`)
...
+
 const fsMachine = Machine({
-  key: "emitFSEvents",
+  key: `emitFSEvents`,
   parallel: true,
-  strict: true,
   states: {
-    CHOKIDAR: {
-      initial: `CHOKIDAR_NOT_READY`,
+    SRC_FS: {
```
Comment (KyleAMathews) [inline packages/gatsby-source-filesystem/src/gatsby-node.js]: Why did you change these names? I'm not sure what your new state names mean...

File: packages/gatsby-source-filesystem/src/gatsby-node.js [hunk 6]
```diff
@@ -11,218 +11,159 @@ const { createId, createFileNode } = require(`./create-file-node`)
...
         },
-        IDLE: {
+        SRC_FS_READY: {
           on: {
-            EMIT_FS_EVENT: `PROCESSING`,
+            SRC_FS_SETUP: `SRC_FS_IDLE`,
           },
         },
-        PROCESSING: {
+        SRC_FS_SETUP: {
```
Comment (KyleAMathews) [inline packages/gatsby-source-filesystem/src/gatsby-node.js]: Please restore the processing state

File: packages/gatsby-source-filesystem/src/gatsby-node.js [hunk 7]
```diff
@@ -11,218 +11,159 @@ const { createId, createFileNode } = require(`./create-file-node`)
...
-      `BOOTSTRAP_FINISHED`
-    )
-  })
-  emitter.on(`TOUCH_NODE`, () => {
-    // If we create a node which is the same as the previous version, createNode
-    // returns TOUCH_NODE and then nothing else happens so we listen to that
-    // to return the state back to IDLE.
-    currentState = fsMachine.transition(currentState.value, `TOUCH_NODE`)
-  })
+  const queue = (() => {
```
Comment (KyleAMathews) [inline packages/gatsby-source-filesystem/src/gatsby-node.js]: I'm not sure what your intention was with this queue?

File: packages/gatsby-source-filesystem/src/create-file-node.js (hunk 3 above)
Comment (ptb) [reply packages/gatsby-source-filesystem/src/create-file-node.js]: Sounds good.

File: packages/gatsby-source-filesystem/src/create-file-node.js (hunk 4 above)
Comment (ptb) [reply packages/gatsby-source-filesystem/src/create-file-node.js]: There very well may be a better way of accomplishing this, but I'm not sure what it would be.

File: packages/gatsby-source-filesystem/src/gatsby-node.js (hunk 6 above)
Comment (ptb) [reply packages/gatsby-source-filesystem/src/gatsby-node.js]: Ok.

File: packages/gatsby-source-filesystem/src/gatsby-node.js (hunk 7 above)
Comment (ptb) [reply packages/gatsby-source-filesystem/src/gatsby-node.js]: Instead of queues for different ready states, this queue is a single unified queue that all chokidar events can add to and if the system is ready to process them, just run them. Having one singleton queue vs multiple and doing the is-ready check inside the queue instead of at each event seems simpler.

File: packages/gatsby-source-filesystem/src/gatsby-node.js (hunk 5 above)
Comment (ptb) [reply packages/gatsby-source-filesystem/src/gatsby-node.js]: I changed them in an attempt to be clearer and simpler. My thinking was that since this is the gatsby-source-filesystem plugin, that SRC_FS would indicate more clearly that this code was being executed inside this plugin vs a more core general purpose file watcher.

You are right though, I certainly don't know your reasons for choosing the names you have. There is a very likely chance that the names cannot/should not be changed. I was trying for the simplest way to achieve (what I believe) is the objective without being excessively verbose.

To decode:
- READY is triggered when chokidar fires it's ready event.
- SETUP is when BOOTSTRAP_FINISHED is fired.
- IDLE is when it is ready but is not processing
- BUSY is when it is processing a file

From my understanding (which granted I am 100% ok with being wrong) ;) there are just a couple events which this plugin needs to be aware of. I'm fairly certain that chokidar's ready event occurs before bootstrap is finished, but I tried to accommodate either order.

The other events that it needs to be concerned about are: is it processing a file currently, or not.

File: packages/gatsby-source-filesystem/src/create-file-node.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (ptb) [reply packages/gatsby-source-filesystem/src/create-file-node.js]: I was hoping to figure out how to make the linked nodes be immediate descendants but couldn't figure out how to without at least one more level below. I don't like edges or nodes and would prefer to eliminate them and change allFile to just files. Direction or help here would be appreciated.

File: packages/gatsby-source-filesystem/src/extend-file-node.js [hunk 8]
```diff
@@ -2,44 +2,36 @@ const { GraphQLString } = require(`graphql`)
...
-      type: GraphQLString,
-      args: {},
-      description: `Copy file to static directory and return public url to it`,
-      resolve: (file, fieldArgs, context) => {
-        const details = getNodeAndSavePathDependency(file.id, context.path)
-        const fileName = `${file.name}-${file.internal.contentDigest}${
-          details.ext
-        }`
+module.exports = ({ getNodeAndSavePathDependency, pathPrefix = ``, type }) =>
+  type.name === `File`
```
[the remark this answers is not in the document]
Comment (ptb) [reply packages/gatsby-source-filesystem/src/extend-file-node.js]: Ok.

**Comment by ptb** [comment]
Just like any other pull request, I'm ok with you not accepting/reverting any of these changes. I used to work at Apple, so my tendency is to pare down code to it's core and try to simplify as much as possible. There are parts of this code that I [probably] don't understand why it was the way it was, so I defer to your judgement. In the end, I needed this functionality for my own projects so if it's not appropriate for a wider audience, I'm ok with that. :)

File: packages/gatsby-source-filesystem/src/create-file-node.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (pieh) [reply packages/gatsby-source-filesystem/src/create-file-node.js]: You can do `files___NODE: ["id-of-first", "id-of-second"]`

File: packages/gatsby-source-filesystem/src/gatsby-node.js (hunk 7 above)
Comment (KyleAMathews) [reply packages/gatsby-source-filesystem/src/gatsby-node.js]: Perhaps this is the source of confusion — there wasn't any queues before — we're using xstate now which let's us create a statechart for the plugin — https://github.com/davidkpiano/xstate/

Comment (ptb) [reply packages/gatsby-source-filesystem/src/gatsby-node.js]: The queues I was referring to are:
https://github.com/gatsbyjs/gatsby/blob/v2/packages/gatsby-source-filesystem/src/gatsby-node.js
lines: 71, 91, 114, 153.

This one function handles all of those cases in a consistent manner. Once chokidar is in a ready state, changes are added to the queue (for no other reason than simplicity and consistency) and immediately flushed.

Comment (KyleAMathews) [reply packages/gatsby-source-filesystem/src/gatsby-node.js]: Ah right :-) There was that queue.

To simplify this PR — could you just back out all the changes to the queue and to the state machine?

Comment (ptb) [reply packages/gatsby-source-filesystem/src/gatsby-node.js]: Done. (I think.)

File: packages/gatsby-source-filesystem/src/create-file-node.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (ptb) [reply packages/gatsby-source-filesystem/src/create-file-node.js]: If I understand you correctly, this code should work?
```
node = {
  ...node,
  files___NODE: stats.isDirectory()
    ? await fs
        .readdir(node.absolutePath)
        .map(file => path.join(node.absolutePath, file))
        .filter(file => fs.statSync(file).isFile())
        .map(file => createNodeId(file))
    : [],
}
```

File: packages/gatsby-source-filesystem/src/create-file-node.js [hunk 9]
```diff
@@ -1,79 +1,74 @@
...
-      birthTime: stats.birthtime,
-      ...slashedFile,
-      ...stats,
-    })
-  )
+  node = {
+    ...node,
+    allFile: {
+      edges: stats.isDirectory()
+        ? await fs
```
[the remark this answers is not in the document]
Comment (ptb) [reply packages/gatsby-source-filesystem/src/create-file-node.js]: Am I correct that there isn't a way to return a value from a function inside an if/else block? (Other than creating a temporary var outside the if and then assigning it to the object after the block?)

File: packages/gatsby-source-filesystem/src/create-file-node.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (pieh) [reply packages/gatsby-source-filesystem/src/create-file-node.js]: Yeah, that's ok

File: packages/gatsby-source-filesystem/src/create-file-node.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (pieh) [reply packages/gatsby-source-filesystem/src/create-file-node.js]:
```js
if (stats.isDirectory()) {
  node.files__NODE = await fs
        .readdir(node.absolutePath)
        .map(file => path.join(node.absolutePath, file))
        .filter(file => fs.statSync(file).isFile())
        .map(file => createNodeId(file))
}
```
I meant this

[the remark this answers is not in the document]
Comment (ptb) [reply packages/gatsby-source-filesystem/src/create-file-node.js]: The auto-linking node magic doesn't work:
```
{
  "data": {
    "allDirectory": {
      "edges": [
        {
          "node": {
            "id": "8b4a7374-f9b9-5fb2-9042-e45aab0d9d91",
            "files__NODE": []
          }
        },
        {
          "node": {
            "id": "e6a79e4a-73c7-567b-9325-255914037bcc",
            "files__NODE": [
              "f54711d3-4270-544b-9479-6de148d4b246"
            ]
          }
        }
      ]
    }
  }
}
```

[the remark this answers is not in the document]
Comment (pieh) [reply packages/gatsby-source-filesystem/src/create-file-node.js]: Oh sorry! My typo mistake - there need to be tripple underscores

[the remark this answers is not in the document]
Comment (ptb) [reply packages/gatsby-source-filesystem/src/create-file-node.js]: It works!

**Comment by m-allanson** [comment]
@pieh @KyleAMathews It looks like @ptb has addressed your comments and this is ready for another look.

File: packages/gatsby-source-filesystem/package.json [hunk 10]
```diff
@@ -30,10 +30,11 @@
     "gatsby-plugin"
   ],
   "license": "MIT",
+  "main": "index.js",
   "repository": "https://github.com/gatsbyjs/gatsby/tree/master/packages/gatsby-source-filesystem",
   "scripts": {
     "build": "babel src --out-dir . --ignore __tests__",
-    "prepublish": "cross-env NODE_ENV=production npm run build",
+    "prepare": "cross-env NODE_ENV=production npm run build",
```
[the remark this answers is not in the document]
Comment (ptb) [reply packages/gatsby-source-filesystem/package.json]: https://docs.npmjs.com/misc/scripts#deprecation-note

“Since npm@1.1.71, the npm CLI has run the prepublish script for both npm publish and npm install, because it's a convenient way to prepare a package for use (some common use cases are described in the section below). It has also turned out to be, in practice, very confusing. As of npm@4.0.0, a new event has been introduced, prepare, that preserves this existing behavior.”

[the remark this answers is not in the document]
Comment (KyleAMathews) [reply packages/gatsby-source-filesystem/package.json]: Could you please remove this? It sounds like a reasonable change but it makes it much harder to review your PRs if you're including multiple unrelated changes.

This is also something, that if we make the change, we'd want to do for all packages.

Could you file an issue about making this change so we can discuss it there?

File: packages/gatsby-source-filesystem/src/create-file-node.js [hunk 11]
```diff
@@ -1,79 +1,72 @@
...
-    const contentDigest = await md5File(slashedFile.absolutePath)
-    const mediaType = mime.getType(slashedFile.ext)
-    internal = {
-      contentDigest,
-      type: `File`,
-      mediaType: mediaType ? mediaType : `application/octet-stream`,
+    node = {
+      ...node,
+      files___NODE: await fs
+        .readdir(node.absolutePath)
```
Comment (pieh) [inline packages/gatsby-source-filesystem/src/create-file-node.js]: Will you be able to change this so we don't use `fs.readdir`? We should have all the needed information comming from chockidar directly, so we shouldn't need to run more I/O operation than needed. This also won't honor default default `ignored` parameter used with chockidar.

**Comment by KyleAMathews** [comment]
Ugh, was deleting stale branches and accidentally deleted this one 🤦‍♂️ Which means I can't restore this PR... @ptb when you have a chance to work on this, please create a new PR and we'll keep working on this! Sorry for the bother!
